java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestDoExchange.java

```diff
@@ -0,0 +1,407 @@
+          Assert.assertEquals(value, iv.get(i));
+          value++;
+        }
+      }
+      Assert.assertEquals(10, value);
+    }
+  }
+
+  /** Emulate a DoGet with a DoExchange. */
+  @Test
+  public void testDoExchangeDoPut() throws Exception {
+    final Schema schema = new Schema(Collections.singletonList(Field.nullable("a", new ArrowType.Int(32, true))));
+    try (final FlightClient.ExchangeReaderWriter stream =
+             client.doExchange(FlightDescriptor.command(EXCHANGE_DO_PUT));
+         final VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator)) {
+      IntVector iv = (IntVector) root.getVector("a");
+      iv.allocateNew();
+
+      stream.getWriter().start(root);
+      int counter = 0;
+      for (int i = 0; i < 10; i++) {
+        for (int row = 0; row < i; row++) {
+          iv.setSafe(row, row);
+        }
+        root.setRowCount(i);
+        counter += i;
+        stream.getWriter().putNext();
+
+        Assert.assertTrue(stream.getReader().next());
+        Assert.assertFalse(stream.getReader().hasRoot());
+        final ArrowBuf metadata = stream.getReader().getLatestMetadata();
+        Assert.assertNotNull(metadata);
+        Assert.assertEquals(counter, metadata.getInt(0));
+      }
+      stream.getWriter().completed();
+
+      while (stream.getReader().next()) {
+        // Drain the stream. Otherwise closing the stream sends a CANCEL which seriously screws with the server.
+        // CANCEL -> runs onCancel handler -> closes the FlightStream early
+      }
+    }
+  }
+
+  /** Test a DoExchange that echoes the client message. */
+  @Test
+  public void testDoExchangeEcho() throws Exception {
+    final Schema schema = new Schema(Collections.singletonList(Field.nullable("a", new ArrowType.Int(32, true))));
+    try (final FlightClient.ExchangeReaderWriter stream = client.doExchange(FlightDescriptor.command(EXCHANGE_ECHO));
+         final VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator)) {
+      final FlightStream reader = stream.getReader();
+
+      // First try writing metadata without starting the Arrow data stream
+      ArrowBuf buf = allocator.buffer(4);
+      buf.writeInt(42);
+      stream.getWriter().putMetadata(buf);
+      buf = allocator.buffer(4);
+      buf.writeInt(84);
+      stream.getWriter().putMetadata(buf);
+
+      Assert.assertTrue(reader.next());
+      Assert.assertFalse(reader.hasRoot());
+      Assert.assertNotNull(reader.getLatestMetadata());
+      Assert.assertEquals(42, reader.getLatestMetadata().getInt(0));
+      Assert.assertTrue(reader.next());
+      Assert.assertFalse(reader.hasRoot());
+      Assert.assertNotNull(reader.getLatestMetadata());
+      Assert.assertEquals(84, reader.getLatestMetadata().getInt(0));
+
+      IntVector iv = (IntVector) root.getVector("a");
+      iv.allocateNew();
+
+      stream.getWriter().start(root);
+      for (int i = 0; i < 10; i++) {
```

Comment:
this seems very complicated for a test, is there a way you could make it clearer what is being tested?